File: leetcode/100-999/leetcode516.cc
#include <iostream>
using namespace std;
class Solution {
public:
    int dp[1005][1005] = {};
    int solve(string& s, int left, int right)
    {
        if (left > right)
            return 0;
        if (dp[left][right])
            return dp[left][right];
        if (left == right)
            return dp[left][right] = 1;
        if (s[left] == s[right])
            return dp[left][right] = 2 + solve(s, left + 1, right - 1);
        // s[left] != s[right]
        return dp[left][right]
            = max(solve(s, left + 1, right), solve(s, left, right - 1));
    }
    int longestPalindromeSubseq(string s)
    {
        return solve(s, 0, s.length() - 1);
    }
};
int main(void) { return 0; }

File: leetcode/100-999/leetcode516.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        # dp[i][j] = from i to j's longest palindrome
        n = len(s)
        dp = [[0] * (n + 1) for _ in range(n + 1)]

        def chaewon(i: int, j: int) -> int:
            if i > j:
                return 0
            if i == j:
                return 1
            if dp[i][j]:
                return dp[i][j]
            if s[i] == s[j]:
                dp[i][j] = 2 + chaewon(i + 1, j - 1)
            else:
                dp[i][j] = max(chaewon(i + 1, j), chaewon(i, j - 1))

            return dp[i][j]

        return chaewon(0, n - 1)
